compiler/rustc_ty_utils/src/assoc.rs

```diff
@@ -162,6 +167,166 @@ fn associated_item_from_impl_item_ref(impl_item_ref: &hir::ImplItemRef) -> ty::A
+            let Some(trait_assoc_id) = tcx.associated_type_for_effects(trait_id) else {
+                return None;
+            };
+
+            // FIXME(effects): span
+            let span = tcx.def_ident_span(def_id).unwrap();
+
+            let impl_assoc_ty = tcx.at(span).create_def(def_id, kw::Empty, DefKind::AssocTy);
+
+            let local_def_id = impl_assoc_ty.def_id();
+            let def_id = local_def_id.to_def_id();
+
+            impl_assoc_ty.feed_hir();
+
+            impl_assoc_ty.def_ident_span(Some(span));
+
+            impl_assoc_ty.associated_item(ty::AssocItem {
+                name: kw::Empty,
+                kind: ty::AssocKind::Type,
+                def_id,
+                trait_item_def_id: Some(trait_assoc_id),
+                container: ty::ImplContainer,
+                fn_has_self_parameter: false,
+                opt_rpitit_info: None,
+                is_effects_desugaring: true,
+            });
+
+            // visibility is public.
+            impl_assoc_ty.visibility(ty::Visibility::Public);
+
+            // no default value.
+            impl_assoc_ty.defaultness(hir::Defaultness::Final);
+
+            // set the type of the associated type! If this is a const impl,
+            // we set to Maybe, otherwise we set to `Runtime`.
+            let type_def_id = if tcx.is_const_trait_impl_raw(impl_def_id.to_def_id()) {
+                tcx.require_lang_item(hir::LangItem::EffectsMaybe, Some(span))
+            } else {
+                tcx.require_lang_item(hir::LangItem::EffectsRuntime, Some(span))
+            };
+            // FIXME(effects): make impls use `Min` for their effect types
+            impl_assoc_ty.type_of(ty::EarlyBinder::bind(Ty::new_adt(
+                tcx,
+                tcx.adt_def(type_def_id),
+                ty::GenericArgs::empty(),
+            )));
+
+            // Copy generics_of of the impl, making the impl as parent.
+            impl_assoc_ty.generics_of({
+                let parent_generics = tcx.generics_of(impl_def_id);
+                let parent_count = parent_generics.parent_count + parent_generics.own_params.len();
+
+                ty::Generics {
+                    parent: Some(impl_def_id.to_def_id()),
+                    parent_count,
+                    own_params: vec![],
+                    param_def_id_to_index: parent_generics.param_def_id_to_index.clone(),
+                    has_self: false,
+                    has_late_bound_regions: None,
+                    host_effect_index: parent_generics.host_effect_index,
+                }
+            });
+
+            // impl_assoc_ty.explicit_item_bounds(ty::EarlyBinder::bind(&[]));
+            impl_assoc_ty.explicit_item_super_predicates(ty::EarlyBinder::bind(&[]));
+
+            // There are no inferred outlives for the synthesized associated type.
+            impl_assoc_ty.inferred_outlives_of(&[]);
+
+            Some(def_id)
```

Comment:
can probably deduplicate a bunch of logic by returning the `Feed` from the match and running all shared feedings after the match.

Reply:
Done